xds/internal/client/xds.go

```diff
@@ -603,32 +604,56 @@ func unmarshalClusterResource(r *anypb.Any, logger *grpclog.PrefixLogger) (strin
 		return "", ClusterUpdate{}, fmt.Errorf("failed to unmarshal resource: %v", err)
 	}
 	logger.Infof("Resource with name: %v, type: %T, contains: %v", cluster.GetName(), cluster, cluster)
-
-	cu, err := validateCluster(cluster)
+	cu, err := validateClusterAndConstructClusterUpdate(cluster)
 	if err != nil {
 		return cluster.GetName(), ClusterUpdate{}, err
 	}
+
 	cu.Raw = r
-	// If the Cluster message in the CDS response did not contain a
-	// serviceName, we will just use the clusterName for EDS.
-	if cu.ServiceName == "" {
-		cu.ServiceName = cluster.GetName()
-	}
+
 	return cluster.GetName(), cu, nil
 }
 
-func validateCluster(cluster *v3clusterpb.Cluster) (ClusterUpdate, error) {
+func validateClusterAndConstructClusterUpdate(cluster *v3clusterpb.Cluster) (ClusterUpdate, error) {
 	emptyUpdate := ClusterUpdate{ServiceName: "", EnableLRS: false}
-	switch {
-	case cluster.GetType() != v3clusterpb.Cluster_EDS:
+	updateToReturn := ClusterUpdate{ServiceName: "", EnableLRS: false}
+	// Only three types of cluster are currently supported by grpc-go.
+	if cluster.GetType() == v3clusterpb.Cluster_EDS {
```

Comment:
`if...else if` can usually be eliminated by moving the code into another function, and return early, instead of `else if`.

In this case, how about adding a new function

```go
func clusterTypeFromCluster(cluster *v3clusterpb.Cluster) (ClusterType, string, []string, error) {
	if cluster.GetType() == v3clusterpb.Cluster_EDS {
		// Process
		return Type_EDS, "service.name", nil, nil
	}

	if cluster.GetType() == DNS {
		// Process
	}

	if type == "envoy.clusters.aggregate" {
		// Do things
		return Type_Aggregate, "", []string{"a", "b"}, nil
    }

	return error("unsupported type")
}
```

and then

```go
func validateClusterAndConstructClusterUpdate(cluster *v3clusterpb.Cluster) (ClusterUpdate, error) {
    typ, serviceName, prioirtizedClusters, err := clusterTypeFromCluster(cluster)
    if err != nil {
		// blahblah
	}

    // handle security fields

    return ClusterUpdate{
		ClusterType: typ,
		ServiceName: serviceName,
		EnableLRS:   cluster.GetLrsServer().GetSelf() != nil,
		SecurityCfg: sc,
		MaxRequests: circuitBreakersFromCluster(cluster),
		PrioritizedClusterNames: prioirtizedClusters,
	}, nil
}
```

Reply:
Interesting. Changed. Perhaps this relates to the concept from Clean Code that I read about keeping functions at one layer of abstraction.

Reply:
Actually, I disagree with this comment. I personally think it's better with an else if (maybe I should've made it a switch to keep it consistent with other parts of the codebase, oh now I remember I couldn't because wasn't switched on same value). I don't like optional return values...perhaps I'm so used to C++ and Java with single return values that I don't like constructing a struct with all these nil values across the different cluster types, I feel like having an else if makes it clear what fields are actually important to the cluster update...whatever these are just musing of an L3. I made the change :P.

Reply:
Oh, no your way is much better, gets rid of this weird empty update thing.

Reply:
Ok, so construct things within a struct literal rather than going struct literal.x = y is what I'm getting from you and Doug's comments.

Reply:
Early returning is usually preferred in go style, as it keeps the code at a minimal indentation. And in most cases, it helps to keep the code clean as well.

Constructing a struct literal at the end is also what I prefer. But I don't remember if it's written in any style guide.

Reply:
Sounds good :). Implicit knowledge transfer!